tools/resourcedocsgen/pkg/registry/svc/metadata.go

```diff
@@ -0,0 +1,256 @@
+	// ListPackageMetadata returns metadata for all packages
+	ListPackageMetadata(ctx context.Context) ([]pkg.PackageMeta, error)
+}
+
+// fileSystemProvider implements PackageMetadataProvider using the local yaml data files
+// in the pulumi/registry repository.
+type fileSystemProvider struct {
+	registryDir string
+}
+
+// registryAPIProvider implements PackageMetadataProvider using the Pulumi API
+// to retrieve package metadata.
+type registryAPIProvider struct {
+	apiURL string
+}
+
+// PackageMetadata represents the API response structure for package metadata
+// from the Pulumi Registry API.
+// TODO: will eventually be available in the "github.com/pulumi/pulumi/sdk/v3/go/common/apitype" package.
+type PackageMetadata struct {
+	Name          string    `json:"name"`
+	Publisher     string    `json:"publisher"`
+	Source        string    `json:"source"`
+	Version       string    `json:"version"`
+	Title         string    `json:"title,omitempty"`
+	Description   string    `json:"description,omitempty"`
+	LogoURL       string    `json:"logoUrl,omitempty"`
+	RepoURL       string    `json:"repoUrl,omitempty"`
+	Category      string    `json:"category,omitempty"`
+	IsFeatured    bool      `json:"isFeatured"`
+	PackageTypes  []string  `json:"packageTypes,omitempty"`
+	PackageStatus string    `json:"packageStatus"`
+	SchemaURL     string    `json:"schemaURL"`
+	CreatedAt     time.Time `json:"createdAt"`
+}
+
+// PackageListResponse represents the API response structure for package lists
+type PackageListResponse struct {
+	Packages []PackageMetadata `json:"packages"`
+}
+
+// NewFileSystemProvider creates a new PackageMetadataProvider that reads from the filesystem
+func NewFileSystemProvider(registryDir string) PackageMetadataProvider {
+	return &fileSystemProvider{
+		registryDir: registryDir,
+	}
+}
+
+// NewAPIProvider creates a new PackageMetadataProvider that reads from the Pulumi API
+func NewAPIProvider(apiURL string) PackageMetadataProvider {
+	return &registryAPIProvider{
+		apiURL: apiURL,
+	}
+}
+
+func getRegistryPackagesPath(repoPath string) string {
+	return filepath.Join(repoPath, "themes", "default", "data", "registry", "packages")
+}
+
+func convertAPIPackageToPackageMeta(apiPkg PackageMetadata) (*pkg.PackageMeta, error) {
+	return &pkg.PackageMeta{
+		Name:          apiPkg.Name,
+		Publisher:     apiPkg.Publisher,
+		Description:   apiPkg.Description,
+		LogoURL:       apiPkg.LogoURL,
+		RepoURL:       apiPkg.RepoURL,
+		Category:      pkg.PackageCategory(apiPkg.Category),
+		Featured:      apiPkg.IsFeatured,
+		Native:        slices.Contains(apiPkg.PackageTypes, "native"),
+		Component:     slices.Contains(apiPkg.PackageTypes, "component"),
+		PackageStatus: pkg.PackageStatus(apiPkg.PackageStatus),
+		SchemaFileURL: apiPkg.SchemaURL,
+		Version:       apiPkg.Version,
+		Title:         apiPkg.Title,
+		UpdatedOn:     apiPkg.CreatedAt.Unix(),
+	}, nil
+}
```

Comment:
This function never errors and always returns a non-null pointer. We can just have it return `pkg.PackageMeta`.
```suggestion
func convertAPIPackageToPackageMeta(apiPkg PackageMetadata) pkg.PackageMeta {
	return pkg.PackageMeta{
		Name:          apiPkg.Name,
		Publisher:     apiPkg.Publisher,
		Description:   apiPkg.Description,
		LogoURL:       apiPkg.LogoURL,
		RepoURL:       apiPkg.RepoURL,
		Category:      pkg.PackageCategory(apiPkg.Category),
		Featured:      apiPkg.IsFeatured,
		Native:        slices.Contains(apiPkg.PackageTypes, "native"),
		Component:     slices.Contains(apiPkg.PackageTypes, "component"),
		PackageStatus: pkg.PackageStatus(apiPkg.PackageStatus),
		SchemaFileURL: apiPkg.SchemaURL,
		Version:       apiPkg.Version,
		Title:         apiPkg.Title,
		UpdatedOn:     apiPkg.CreatedAt.Unix(),
	}
}
```